plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py

```diff
@@ -491,27 +488,37 @@ def make_template(in_str):
 
             # When using dbt-templater, trailing newlines are ALWAYS REMOVED during
             # compiling. Unless fixed (like below), this will cause:
-            #    1. L009 linting errors when running "sqlfluff lint foo_bar.sql"
+            #    1. Assertion errors in TemplatedFile, when it sanity checks the
+            #       contents of the sliced_file array.
+            #    2. L009 linting errors when running "sqlfluff lint foo_bar.sql"
             #       since the linter will use the compiled code with the newlines
             #       removed.
-            #    2. "No newline at end of file" warnings in Git/GitHub since
+            #    3. "No newline at end of file" warnings in Git/GitHub since
             #       sqlfluff uses the compiled SQL to write fixes back to the
             #       source SQL in the dbt model.
-            # The solution is:
+            #
+            # The solution is (note that both the raw and compiled files have
+            # had trailing newline(s) removed by the dbt-templater.
             #    1. Check for trailing newlines before compiling by looking at the
-            #       raw SQL in the source dbt file, store the count of trailing
+            #       raw SQL in the source dbt file. Remember the count of trailing
             #       newlines.
-            #    2. Append the count from #1 above to the node.raw_sql and
-            #       compiled_sql objects, both of which have had the trailing
-            #       newlines removed by the dbt-templater.
-            node.raw_sql = node.raw_sql + "\n" * n_trailing_newlines
+            #    2. Set node.raw_sql to the original source file contents.
+            #    3. Append the count from #1 above to compiled_sql. (In
+            #       production, slice_file() does not usually use this string,
+            #       but some test scenarios do.
+            #    4. If there was at least one trailing newline, have
+            #       slice_file() append a newline after rendering the template.
+            #       (Note that slice_file() usually renders the template itself,
+            #       rather than using compiled_sql).
+            node.raw_sql = source_dbt_sql
             compiled_sql = compiled_sql + "\n" * n_trailing_newlines
 
             raw_sliced, sliced_file, templated_sql = self.slice_file(
                 source_dbt_sql,
                 compiled_sql,
                 config=config,
                 make_template=make_template,
+                append_to_templated="\n" if n_trailing_newlines else "",
```

Comment:
Why only appending 1 line, instead of all of them? Guess it doesn't really matter as the newlines aren't needed, but still might be cleaner to put them all back. Also would we need then if implement #2634?

Reply:
This was the bug. The old code was based on a mistaken belief that Jinja & dbt were removing **all** the trailing blank lines, but it was actually only removing the **last** one. We only noticed this now, because the internal checks I added to the `TemplatedFile` constructor caught the inconsistency. None of our existing tests had multiple trailing newlines (I added one). The old code was basically _doubling_ the number of trailing newlines. Oops!

Reply:
This isn't clear. Can we add more explanation to lines 509-510 to explain this?

Reply:
Will do.